inputremapper/gui/data_manager.py

```diff
@@ -0,0 +1,446 @@
+#!/usr/bin/python3
+# -*- coding: utf-8 -*-
+# input-remapper - GUI for device specific keyboard mappings
+# Copyright (C) 2022 sezanzeb <proxima@sezanzeb.de>
+#
+# This file is part of input-remapper.
+#
+# input-remapper is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# input-remapper is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with input-remapper.  If not, see <https://www.gnu.org/licenses/>.
+import glob
+import os
+import re
+import time
+from typing import Optional, List, Tuple
+
+from inputremapper.configs.global_config import GlobalConfig
+from inputremapper.configs.mapping import UIMapping, MappingData
+from inputremapper.configs.preset import Preset
+from inputremapper.configs.paths import get_preset_path, mkdir, split_all
+from inputremapper.daemon import DaemonProxy
+from inputremapper.event_combination import EventCombination
+from inputremapper.exceptions import DataManagementError
+from inputremapper.groups import _Group
+from inputremapper.gui.data_bus import (
+    DataBus,
+    GroupData,
+    PresetData,
+    StatusData,
+    CombinationUpdate,
+    UInputsData,
+)
+from inputremapper.gui.reader import Reader
+from inputremapper.gui.utils import CTX_MAPPING, CTX_APPLY
+from inputremapper.gui.gettext import _
+from inputremapper.injection.global_uinputs import GlobalUInputs
+from inputremapper.logger import logger
+
+DEFAULT_PRESET_NAME = "new preset"
+
+# useful type aliases
+Name = str
+GroupKey = str
+
+
+class DataManager:
+    """provides an interface to create configurations
+    and send commands to the service"""
```

Comment:
Can/shouldn't this be separated into one class for the config, and one for the communication?

Reply:
Kinda, but not really. At first I had it separated into 3 classes: config, reader and daemon but I think that is not a good idea because I might end up integrating the helper in the daemon (the helper already uses the new event pipeline to read events).

I tried to combine reader and daemon in a communication class. 
But this communication class would be just a thin wrapper around the reader and the daemon without any logic itself.

The config needs to know about the reader in order to find all groups, the daemon needs to know about the config, so it can start injecting the correct preset. So maybe we can have them separated but they would still share each other as a object.

Having them in on class has the advantage that all the state of the application is in one place so it can act as a interface if someone else decides to create a UI.

Reply:
> integrating the helper in the daemon

the reason why they are separate is, because no user process should be able to ask the daemon for input events at all times. Only during the lifetime of the gui, which is also the lifetime of the helper, is reading keys permitted for the user processes.

Maybe there is an alternative solution with a pipe that is only readable from by elevated processes. But since the gui can't run as root in wayland a helper process would still be needed

Reply:
yeah this is nothing I will do anytime soon. And the daemon is currently not designed to do this easily. But it would open a lot of possibilities.